src/armature/Silhouette.ts

```diff
@@ -0,0 +1,209 @@
+                addedGeometry.map((g: GeometryNode) => {
+                    if (this.positionBuffers.get(g.geometry) === undefined) {
+                        this.positionBuffers.set(g.geometry, this.regl.buffer(g.geometry.vertices));
+                    }
+                    if (this.indicesBuffers.get(g.geometry) === undefined) {
+                        this.indicesBuffers.set(g.geometry, this.regl.elements(g.geometry.indices));
+                    }
+
+                    return {
+                        model: g.localToGlobalTransform(),
+                        projection: this.projection,
+                        positions: this.positionBuffers.get(g.geometry),
+                        indices: this.indicesBuffers.get(g.geometry)
+                    };
+                })
+            );
+
+            this.regl.read(data);
+        });
+
+        const lastNodeInParent = instance.getModel().nodes[startIndex - 1];
+        const parentImage = this.lastImage.get(lastNodeInParent);
+
+        let difference = 0;
+        let costViz: Uint8ClampedArray | undefined;
+        if (this.debug) {
+            costViz = new Uint8ClampedArray(this.target.width * this.target.height * 4);
+        }
+        for (let x = 0; x < this.target.width; x += 1) {
+            for (let y = 0; y < this.target.height; y += 1) {
+                const index = y * this.target.width * 4 + x * 4;
+
+                if (parentImage !== undefined) {
+                    data[index] = Math.min(data[index], parentImage[index]);
+                }
+
+                if (costViz !== undefined) {
+                    costViz[index] = 255;
+                    costViz[index + 1] = 255;
+                    costViz[index + 2] = 255;
+                    costViz[index + 3] = 255;
+                }
+
+                const incrementalDifference = this.target.data[index] - data[index];
+
+                // Only have a penalty for not going into a black area of the target;
+                // SOSMC should help ensure we don't go into the white region since
+                // it adds no benefit
+                if (incrementalDifference < 0) {
+                    difference -= incrementalDifference;
+
+                    if (costViz !== undefined) {
+                        costViz[index + 1] = incrementalDifference + 255;
+                        costViz[index + 2] = incrementalDifference + 255;
+                    }
+                }
+            }
+        }
+
+        this.lastImage.set(instance.getModel().nodes[instance.getModel().nodes.length - 1], data);
+
+        if (costViz !== undefined) {
+            this.ctx.putImageData(
+                new ImageData(costViz, this.target.width, this.target.height),
+                0,
+                0
+            );
+        }
+
+        const normalizedDifference = difference / (this.target.width * this.target.height * 255);
+
+        return { realCost: normalizedDifference * 1200 - 150, heuristicCost: 0 };
```

Comment:
Just curious, where do the 1200 and 150 come from here?

Reply:
Basically just experimenting with finding a good range and offset for the costs so that when we do `Math.exp()` to get weightings, it doesn't get overwhelmed by things that are initially promising but go nowhere (e.g. single-branch trees) and also lets good things surface without too much noise. There's nothing special about these numbers, but they happen to work OK for the sample inputs (because the maximum cost depends on the size of the sample input that is a target region.) This was super finicky though.